File: general_pc/cancel/1.일반취소/Java/PartialRefund/partialRefund.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Map;

import org.json.simple.JSONObject;

public class partialRefund {

	public static void main(String[] args) throws Exception {

		SHA512 sha512 = new SHA512();
		Date date_now = new Date(System.currentTimeMillis());
		SimpleDateFormat fourteen_format = new SimpleDateFormat("yyyyMMddHHmmss");

		//step1. 요청을 위한 파라미터 설정						
		String key = "ItEQKi3rY7uvDS8l"; 											
		String mid = "INIpayTest";	
		String type = "partialRefund";
		String timestamp = fourteen_format.format(date_now);		
		String clientIp = "127.0.0.1";
		
		
		Map<String, Object> data1 = new HashMap<String, Object>();
	 	data1.put("tid", "");
	 	data1.put("msg", "부분취소 테스트");
	 	data1.put("price", "700");
	 	data1.put("confirmPrice", "300");
	 	data1.put("currency", "WON");
	 	data1.put("tax", "0");
	 	data1.put("taxfree", "0");
	 	
	 	JSONObject data = new JSONObject(data1);
	 	

		// Hash Encryption 
		String plainTxt = key + mid + type + timestamp + data ;
		plainTxt = plainTxt.replaceAll("\\\\", "");
		String hashData = sha512.hash(plainTxt);
		
		
		// reqeust URL
		String apiUrl = "https://iniapi.inicis.com/v2/pg/partialRefund";
		
		 JSONObject respJson = new JSONObject();
		    respJson.put("mid", mid);
		    respJson.put("type", type);
		    respJson.put("timestamp",timestamp);
		    respJson.put("clientIp",clientIp);
		    respJson.put("data",data);
		    respJson.put("hashData",hashData);
		   
		
		//step2. key=value 로 post 요청
		try {
			URL reqUrl = new URL(apiUrl);
			HttpURLConnection conn = (HttpURLConnection) reqUrl.openConnection();
			
			if (conn != null) {
				conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
				conn.setRequestMethod("POST");
				conn.setDefaultUseCaches(false);
				conn.setDoOutput(true);
				
				if (conn.getDoOutput()) {
					conn.getOutputStream().write(respJson.toString().getBytes("UTF-8"));
					conn.getOutputStream().flush();
					conn.getOutputStream().close();
				}

				conn.connect();
				
					BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
					
					//step3. 요청 결과
					System.out.println(br.readLine()); 
					br.close();
				}

		}catch(Exception e ) {
			e.printStackTrace();
		} 
	}
}

File: general_pc/cancel/1.일반취소/node/PartialRefund/PartialRefund.js
const express = require("express"); 
const PartialRefund = express();
const crypto = require('crypto'); 
const bodyParser = require("body-parser");
const request = require('request');

PartialRefund.use(bodyParser.json());
PartialRefund.use(bodyParser.urlencoded({extended : true}));

Date.prototype.YYYYMMDDHHMMSS = function () {
    var yyyy = this.getFullYear().toString();
    var MM = pad(this.getMonth() + 1,2);
    var dd = pad(this.getDate(), 2);
    var hh = pad(this.getHours(), 2);
    var mm = pad(this.getMinutes(), 2)
    var ss = pad(this.getSeconds(), 2)

    return yyyy +  MM + dd+  hh + mm + ss;
};

function pad(number, length) {
    var str = '' + number;
    while (str.length < length) {
      str = '0' + str;
    }
    return str;
}

var nowDate = new Date();
var HashMap = require ('hashmap');
    

//step1. 요청을 위한 파라미터 설정
const key = "ItEQKi3rY7uvDS8l";               // INIAPI Key
const mid = "INIpayTest";                     // 상점ID
const type = "partialRefund";
const timestamp = nowDate.YYYYMMDDHHMMSS();
const clientIp = "192.0.0.1";

const data = {
    tid : "StdpayCARDINIpayTest20231222110519706364",
    msg : "테스트 부분취소",
    price : "500",                           // 부분취소 요청금액
    confirmPrice : "0"                       // 부분취소 후 남은금액
};

// Hash Encryption
let plainTxt = key + mid + type + timestamp + JSON.stringify(data);
console.log("PLAINTXT : " + plainTxt)
plainTxt = plainTxt.replace(/\\/g, '');
const hashData = crypto.createHash('sha512').update(plainTxt).digest('hex');

console.log("HASHDATA : " + hashData);


//step2. API 요청/응답
const apiUrl = "https://iniapi.inicis.com/v2/pg/partialRefund"

let prams = {
    mid : mid,
    type : type,
    timestamp : timestamp,
    clientIp : clientIp, 
    data : data,
    hashData : hashData
}

let prams1 = JSON.stringify(prams)

let prams2 = prams1.replace(/\\/g, '');
console.log("REQUEST DATA : " + prams2);

request.post({ method: 'POST', uri: apiUrl, body: prams, json: true}, (err,httpResponse,body) =>{ 
    console.log("RESPONSE DATA :",body)
});